src/main/java/org/rocstreaming/roctoolkit/Endpoint.java

```diff
@@ -0,0 +1,187 @@
+ * resource. All these parts together are unambiguously represented
+ * by a URI. The user may set or get the entire URI or its individual parts.
+ * </p>
+ * <h3>Endpoint URI</h3>
+ * <p>
+ * Endpoint URI syntax is a subset of the syntax defined in RFC 3986:
+ * <p>
+ * protocol://host[:port][/path][?query]
+ * <p>
+ * Examples:
+ * <ul>
+ *     <li>"rtsp://localhost:123/path?query"</li>
+ *     <li>"rtp+rs8m://localhost:123"</li>
+ *     <li>"rtp://127.0.0.1:123"</li>
+ *     <li>"rtp://[::1]:123"</li>
+ * </ul>
+ * <p>
+ * The following protocols (schemes) are supported:
+ * <ul>
+ *     <li> "rtp://"       ({@link Protocol#RTP RTP})</li>
+ *     <li> "rtp+rs8m://"  ({@link Protocol#RTP_RS8M_SOURCE RTP_RS8M_SOURCE})</li>
+ *     <li> "rs8m://"      ({@link Protocol#RS8M_REPAIR RS8M_REPAIR})</li>
+ *     <li> "rtp+ldpc://"  ({@link Protocol#RTP_LDPC_SOURCE RTP_LDPC_SOURCE})</li>
+ *     <li> "ldpc://"      ({@link Protocol#LDPC_REPAIR LDPC_REPAIR})</li>
+ * </ul>
+ * <p>
+ * The host field should be either FQDN (domain name), or IPv4 address, or
+ * IPv6 address in square brackets.
+ * <p>
+ * The port field can be omitted if the protocol defines standard port. Otherwise,
+ * the port can not be omitted. For example, RTSP defines standard port,
+ * but RTP doesn't.
+ * <p>
+ * The path and query fields are allowed only for protocols that support them.
+ * For example, they're supported by RTSP, but not by RTP.
+ * <p>
+ * <h3>Thread-safety</h3>
+ * <p>
+ * Can't be used concurrently
+ */
+public class Endpoint {
+
+    private Protocol protocol;
+
+    private String host;
+
+    private int port;
+
+    private String resource;
+
+    /**
+     * Create endpoint from uri
+     *
+     * @param uri uri
+     */
+    public Endpoint(String uri) {
+        init(uri);
+    }
+
+    /**
+     * Create endpoint from components
+     *
+     * @param protocol protocol
+     * @param host     host specifies FQDN, IPv4 address, or IPv6 address
+     * @param port     port specifies UDP or TCP port in range [0; 65535]
+     *                 <p>
+     *                 When binding an endpoint, the port may be set to zero to select a random port.
+     *                 The selected port will be then written back to the endpoint. When connecting
+     *                 an endpoint, the port should be positive.
+     *                 <p>
+     *                 If port is set to -1, the standard port for endpoint protocol is used. This is
+     *                 allowed only if the protocol defines its standard port.
+     * @param resource resource nullable. Specifies percent-encoded path and query
+     */
+    public Endpoint(Protocol protocol, String host, int port, String resource) {
+        if (protocol == null || host == null || host.isEmpty()) throw new IllegalArgumentException();
+        this.protocol = protocol;
+        this.host = host;
+        this.port = port;
+        this.resource = resource;
+    }
```

Comment:
I think we should also validate endpoint using C api because not all combinations are allowed. E.g., resource is not allowed for some protocols, or port can be -1 only for certain protocols.

By validation, I mean building endpoint from parts and try to call get_uri.